### PR TITLE
Avoid race conditions in DNSSnooper's cached domains

### LINKED ISSUE #2636: Race condition in DNSSnooper cached domains
The Go runtime stated panicking after we upgraded to go 1.8 at #2621 

```
fatal error: concurrent map iteration and map write

goroutine 2771 [running]:
runtime.throw(0x17a763c, 0x26)
    /usr/local/go/src/runtime/panic.go:596 +0x95 fp=0xc423561240 sp=0xc423561220
runtime.mapiternext(0xc4235613d8)
    /usr/local/go/src/runtime/hashmap.go:737 +0x7ee fp=0xc4235612f0 sp=0xc423561240
runtime.mapiterinit(0x1579ce0, 0xc4219fe450, 0xc4235613d8)
    /usr/local/go/src/runtime/hashmap.go:727 +0x2b3 fp=0xc423561348 sp=0xc4235612f0
github.com/weaveworks/scope/probe/endpoint.(*DNSSnooper).CachedNamesForIP(0xc4218e40f0, 0xc424c14744, 0xc, 0xc4268cace0, 0x10, 0x0)
    /go/src/github.com/weaveworks/scope/probe/endpoint/dns_snooper_linux_amd64.go:105 +0x136 fp=0xc423561448 sp=0xc423561348
github.com/weaveworks/scope/probe/endpoint.(*connectionTracker).makeEndpointNode(0xc42000ab60, 0xc424c14750, 0xa, 0xc424c14744, 0xc, 0x50, 0xc427b48ab0, 0x0, 0x0, 0x0, ...)
    /go/src/github.com/weaveworks/scope/probe/endpoint/connection_tracker.go:227 +0x13d fp=0xc423561728 sp=0xc423561448
github.com/weaveworks/scope/probe/endpoint.(*connectionTracker).addConnection(0xc42000ab60, 0xc4263e4580, 0xc424c14730, 0xe, 0xc424c14744, 0xc, 0x50a9d4, 0xc424c14750, 0xa, 0xc427b48a80, ...)
    /go/src/github.com/weaveworks/scope/probe/endpoint/connection_tracker.go:218 +0x118 fp=0xc423561ca0 sp=0xc423561728
github.com/weaveworks/scope/probe/endpoint.(*connectionTracker).performEbpfTrack.func1(0xc424c14730, 0xe, 0xc424c14744, 0xc, 0x50a9d4, 0xc424c14750, 0xa, 0x0, 0x1dd9)
    /go/src/github.com/weaveworks/scope/probe/endpoint/connection_tracker.go:208 +0x406 fp=0xc423561de0 sp=0xc423561ca0
github.com/weaveworks/scope/probe/endpoint.(*EbpfTracker).walkConnections(0xc421cd0940, 0xc426d83320)
    /go/src/github.com/weaveworks/scope/probe/endpoint/ebpf.go:253 +0x257 fp=0xc423561fd0 sp=0xc423561de0
github.com/weaveworks/scope/probe/endpoint.(*connectionTracker).performEbpfTrack(0xc42000ab60, 0xc4263e4580, 0xc4246d4060, 0x15, 0x0, 0x0)
    /go/src/github.com/weaveworks/scope/probe/endpoint/connection_tracker.go:211 +0xae fp=0xc423561ff8 sp=0xc423561fd0
github.com/weaveworks/scope/probe/endpoint.(*connectionTracker).ReportConnections(0xc42000ab60, 0xc4263e4580)
    /go/src/github.com/weaveworks/scope/probe/endpoint/connection_tracker.go:87 +0x1ea fp=0xc423562058 sp=0xc423561ff8
github.com/weaveworks/scope/probe/endpoint.(*Reporter).Report(0xc42000ab00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /go/src/github.com/weaveworks/scope/probe/endpoint/reporter.go:100 +0x121 fp=0xc423562fc0 sp=0xc423562058
github.com/weaveworks/scope/probe.(*Probe).report.func1(0xc4204bd5f0, 0xc4224ff980, 0x2467980, 0xc42000ab00)
    /go/src/github.com/weaveworks/scope/probe/probe.go:155 +0x102 fp=0xc423563fc0 sp=0xc423562fc0
runtime.goexit()
    /usr/local/go/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc423563fc8 sp=0xc423563fc0
created by github.com/weaveworks/scope/probe.(*Probe).report
    /go/src/github.com/weaveworks/scope/probe/probe.go:165 +0xec
```

### PR DESCRIPTION
**Pull request by 2opremio**
Fixes #2636

### CODE REVIEW

File: probe/endpoint/dns_snooper_linux_amd64.go [hunk 1]
```diff
@@ -29,6 +30,14 @@ type DNSSnooper struct {
 	decodingErrorCounts map[string]uint64 // for limiting
 }
 
+type domainSet struct {
+	// Not worth using a RMutex since we don't expect a lot of contention
+	// and they are considerably larger (8 bytes vs 24 bytes), which would
+	// bloat the cache
+	sync.Mutex
```
Comment (rade) [inline probe/endpoint/dns_snooper_linux_amd64.go]: Shouldn't this be a pointer? " A Mutex must not be copied after first use."

Comment (2opremio) [inline probe/endpoint/dns_snooper_linux_amd64.go]: Very true, I initially created a global lock in the `DNSSnooper` (which is passed by pointer) and overlooked that detail when moving it here.

However, after thinking about I think it's fine (gcache doesn't copy, it uses interfaces, which hold a pointer).

**Comment by 2opremio** [comment]
> Do we really want such fine-grained locking? Rather than a single lock.

I started with a global lock, but I figured there would be some contention and locks are small (8 bytes).

File: probe/endpoint/dns_snooper_linux_amd64.go (hunk 1 above)
Comment (rade) [inline probe/endpoint/dns_snooper_linux_amd64.go]: It's not gcache that you need to worry about but the map entries. Which certainly *do* hold structs, not pointers.

Comment (rade) [inline probe/endpoint/dns_snooper_linux_amd64.go]: Oh, ignore me. I'm wrong.

Comment (rade) [inline probe/endpoint/dns_snooper_linux_amd64.go]: Actually, I think *you* are wrong: https://play.golang.org/p/C1cUBezCpy

File: probe/endpoint/dns_snooper_linux_amd64.go [hunk 2]
```diff
@@ -29,6 +30,14 @@ type DNSSnooper struct {
 	decodingErrorCounts map[string]uint64 // for limiting
 }
 
+type domainSet struct {
+	// Not worth using a RMutex since we don't expect a lot of contention
+	// and they are considerably larger (8 bytes vs 24 bytes), which would
+	// bloat the cache
+	mutex   *sync.Mutex
+	domains map[string]struct{}
+}
```
Comment (rade) [inline probe/endpoint/dns_snooper_linux_amd64.go]: I think it would be cleaner to store `*domainSets`, and keep the mutex as an embedded struct.

**Comment by 2opremio** [comment]
Alright, I caved in (midsommar-good-mood :) ) and modified the code to use a global mutex instead.

File: probe/endpoint/dns_snooper_linux_amd64.go [hunk 3]
```diff
@@ -23,8 +24,11 @@ const (
 
 // DNSSnooper is a snopper of DNS queries
 type DNSSnooper struct {
-	stop                chan struct{}
-	pcapHandle          *pcap.Handle
+	stop       chan struct{}
+	pcapHandle *pcap.Handle
+	// gcache is goroutine-safe, but the
+	// values cached values aren't
```
Comment (rade) [inline probe/endpoint/dns_snooper_linux_amd64.go]: typo.

**Pull request merged**